### ISSUE TITLE
Failed to build Rider Link

### ISSUE THREAD

**Issue by shamsfk** [JetBrains/UnrealLink#23]
I get RiderLink error both in editor and project

Failed to build RiderLink plugin
			Check build logs for more info
			Help > Diagnostic Tools > Show Log in Explorer

Log:
[idea.log](https://github.com/JetBrains/UnrealLink/files/4680620/idea.log)

**Comment by DecoyRS** [comment]
Hi @shamsfk 
Could you please share the whole log folder? I'm really interested in backend.log and backend-err.log, there should be info from the building process and reason why it couldn't find your UE

**Comment by afonsocsc** [comment]
I'm having the same issue, I've also tried with a new blank c++ Unreal project.
I have Unreal 4.24.3 installed in "C:\develop\UE_4.24"
Rider is installed in "D:\apps\JetBrains\Rider for Unreal Engine 2020.1.2"

Log files:
[backend-err.log](https://github.com/JetBrains/UnrealLink/files/4681876/backend-err.log)
[backend.log](https://github.com/JetBrains/UnrealLink/files/4681874/backend.log)

**Comment by shamsfk** [comment]
[backend.log](https://github.com/JetBrains/UnrealLink/files/4682403/backend.log)
[backend-err.log](https://github.com/JetBrains/UnrealLink/files/4682411/backend-err.log)

I'm using 4.25.0, the only unusual thing I can think of - it is installed on drive D to save space on system ssd

**Comment by DecoyRS** [comment]
@afonsocsc @shamsfk can you share your {GameName}.uproject files?
**Short version**
Check if you have these registry keys:
HKEY_LOCAL_MACHINE\SOFTWARE\EpicGames\Unreal Engine\4.25
or
HKEY_CURRENT_USER\SOFTWARE\EpicGames\Unreal Engine\4.25
and they point to a valid Unreal Engine root folder and this folder has
`{UnrealEngineRoot}/Engine/Binaries/DotNET/UnrealBuildTool.exe`

**Long version**
How Rider's trying to find Unreal Engine root:
1. Look into uproject file and take "EngineVersion" field
2. If it's empty, look into parent folder recursively and check if it's UE
3. If it has actual engine version (e.g. "4.25"), it will look into Windows registry, fields
HKEY_LOCAL_MACHINE\SOFTWARE\EpicGames\Unreal Engine\{EngineVersion}
or
HKEY_CURRENT_USER\SOFTWARE\EpicGames\Unreal Engine\{EngineVersion}
4. If it has GUID of custom engine, it will look into Windows registry, fields:
HKEY_LOCAL_MACHINE\Software\Epic Games\Unreal Engine\Builds , field {GUID}
or
HKEY_CURRENT_USER\Software\Epic Games\Unreal Engine\Builds , field {GUID}

And how does it determine that it's an Unreal Engine root:
it checks if folder has 
{UnrealEngineRoot}/Engine/Binaries/DotNET/UnrealBuildTool.exe

**Comment by shamsfk** [comment]
[Test.zip](https://github.com/JetBrains/UnrealLink/files/4683100/Test.zip)
![Annotation 2020-05-26 180801](https://user-images.githubusercontent.com/4471466/82917331-f77a4280-9f7b-11ea-863a-cd81b467cd2e.png)

project file is in the zip archive.

I don't have 4.25 folder at the location at all (see screenshot) (this is in the CURRENT_USER path, LOCAL_MACHINE path has no EpicGames)
UE was installed using Epic Games Launcher, nothing was customized with exception to its location being set to D drive

**Comment by shamsfk** [comment]
nor I have any entries named 4.25
![Annotation 2020-05-26 181613](https://user-images.githubusercontent.com/4471466/82918210-0f9e9180-9f7d-11ea-9ce7-ca5b0ded0374.png)

**Comment by DecoyRS** [comment]
Thanks for the info @shamsfk !
For Engines from EGS, "EpicGames" doesn't have spaces in it. Can you double check it one more time, please? And if it won't find it, can you run the full search for the path of your local install of UE.
Example:
![image](https://user-images.githubusercontent.com/1694911/82926253-96f10280-9f87-11ea-8e9b-061d55f63954.png)
In my case I have UE 4.25 installed to E:\GAMES\EpicGames\UE_4.25 and I have selected the root node in Registry Editor.

**Comment by shamsfk** [comment]
Mine has space, I did a full registry search by the path, it did find some Visual Assist X, Firewall, and other unrelated stuff, but no EpicGames with no space. Mind that engine is fully working, everything is compiling, running, and stuff, just not the RiderLink.

**Comment by SosoVG** [comment]
I was experiencing the same issue (the plugin couldn’t be installed either in-game or in-engine). In the Registry Editor I had keys for other engine versions, but not for the 4.25 (the one I was trying to install the plugin for).

After adding the key manually with the _InstalledDirectoy_ pointing to my UE installation folder. It did get installed correctly.

I added the key to the HKEY_LOCAL_MACHINE\SOFTWARE\EpicGames\Unreal Engine field.

**Comment by DecoyRS** [comment]
@shamsfk 
It has space for Engine that is built from source (github) and after you've run Setup.bat.
For engines installed from EGS, it will always lack a space in the name.
Looks like it's a bug in UE 4.25 installation, although I do have it in my registry.

**Comment by shamsfk** [comment]
I absolutely did not compile it. 3 days ago I downloaded EGL and installed UE4.25 that's it. There is no EpicGames without space in the registry at all. With it - there is, but it has no versions inside.

**Comment by DecoyRS** [comment]
I fully understand that. It means, that for whatever reason, installation of UE 4.25 has bugged out and didn't register itself, so there's no possible way for me to find the installation folder.

**Comment by afonsocsc** [comment]
@DecoyRS thank you!
On the registry, I only had an entry for 4.25 and not for 4.24. I've manually added the 4.24 entry and the UnrealLink plugin installed correctly.

**Comment by shamsfk** [comment]
VAX somehow found it. It has the right path in its registry.

**Comment by DecoyRS** [comment]
VAX simply checks if solution has a UE4 project and considers it as an Unreal Engine root.
I couldn't do that because we have use-cases where we don't have an Engine inside solution, but we do have a uproject file on our hand.
I've found another way to find install locations for Unreal Engine's, it should be available in the next version of Rider 4UE and will fix your issues.
I'll close this issue when new version of Rider will be released. Meanwhile, a workaround is to add a record to Windows Registry.

**Comment by Dolichomps** [comment]
I had the exact same issue as @shamsfk, no EpicGames without space in the registry with my UE 4.25 install. I decided to verify my installation of the engine with the Epic Games Launcher and upon completion it had registered itself and the RiderLink plugin installed properly. This might help while waiting for the next version of Rider 4UE.

**Comment by OrzGame** [comment]
I'm having the same issue on my laptop,however it works fine on my desktop.

**Comment by jhuhtanen** [comment]
I had clean desktop installed on Win10 and whenever I tried to open c++ project with rider it complained about RiderLink missing and needs to install it. Install always failed with error being "Message = [UnrealLink]: Failed to build plugin for C:\Projects\UE4\XXXX\YYYYY.uproject, couldn't find Unreal Engine root" (this from rider.log)

This is how I got mine to work: I added keys to registry under "Computer\HKEY_LOCAL_MACHINE\SOFTWARE" for my current engine version (4.25) like so: "Computer\HKEY_LOCAL_MACHINE\SOFTWARE\EpicGames\Unreal Engine\4.25" and new string value for "InstalledDirectory" with "C:\Program Files\Epic Games\UE_4.25" where my engine root is.

I didn't have "EpicGames" under "Computer\HKEY_LOCAL_MACHINE\SOFTWARE" instead I had keys under "Computer\HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node". I tried to add the missing keys under WOW6432Node, but that does not work. Only adding keys under "Computer\HKEY_LOCAL_MACHINE\SOFTWARE\EpicGame" did the trick. Hope this helps others!

**Comment by DecoyRS** [comment]
Days since win registry bullshit has bite my ass - 0. Thank you @jhuhtanen for this info, I'll add check for explicit 64 and 32 registry keys.

**Comment by RVillani** [comment]
I have the registry keys set correctly, but it still fails. :(
I've attached all my logs.
[backend.log](https://github.com/JetBrains/UnrealLink/files/4814990/backend.log)
[backend-err.log](https://github.com/JetBrains/UnrealLink/files/4814992/backend-err.log)
[backend-out.log](https://github.com/JetBrains/UnrealLink/files/4814993/backend-out.log)
[backend-protocol.log](https://github.com/JetBrains/UnrealLink/files/4814994/backend-protocol.log)
[frontend.log](https://github.com/JetBrains/UnrealLink/files/4814995/frontend.log)

**Comment by DecoyRS** [comment]
Hello @RVillani !
We've discovered recently that latest versions of VS19 are failing to build RiderLink due to RD being build with VS17 toolchain. They were backward compatible before, but it's not true anymore(
We've fixed it in the latest version of UnrealLink/RiderLink, it'll be available for the users in the next couple of weeks.

**Comment by RVillani** [comment]
I think I know what VS update created this incompatibility issue. They mentioned something on the third item of [these release notes](https://docs.microsoft.com/en-us/visualstudio/releases/2019/release-notes#top-issues-fixed-in-visual-studio-2019-version-1661).
But I'm trying to build the plugin from RD UE4. Doesn't it use its own version of MSVC?

And if I can't build it anyway, is there anywhere where I can download the built binaries for UE 4.24?

Thanks for your help!

**Comment by hikingyoung** [comment]
Hi @DecoyRS 
I have the same problem. As I have read all mentioned above, I uninstall VS2019 and install VS2017 instead, but nothing changed, still "Failed to build RiderLink plugin" 
1. The register is OK cuz' I have checked it very carefully, the UE4.25 path is clear and right.
2. UE4-->Editor-->SourceCode is selected "Rider" already.
I really hope this can be fixed ASAP cuz I love your IDE so much that I just can't wait for  that "couple of weeks" long. 
Here is my log folder if you need it.
[log.zip](https://github.com/JetBrains/UnrealLink/files/4817977/log.zip)
Thanks

**Comment by DecoyRS** [comment]
Hi @RVillani 
> Doesn't it use its own version of MSVC?

No, they use either bundled with VS version, or with MSBuild toolchain.

Taking latest version of pre-built RD is somewhat complicated.
You'll need to:
* take latest https://github.com/JetBrains/UnrealLink/tree/net202/src/cpp/RiderLink/Source/RD folder, it has only RD.Build.cs
* take JetBrains.RdFramework.Cpp.201.1.56-win-vs17.nupkg and JetBrains.RdFramework.Cpp.201.1.56-win-vs19.nupkg from
https://bintray.com/jetbrains/rd-nuget/JetBrains.RdFramework.Cpp/view/files?sort=name&order=asc#files/
* unzip nupkg files and take contents of their subfolder called `native` and copy them to `RD` respectively, e.g.
JetBrains.RdFramework.Cpp.201.1.56-win-vs17.nupkg/native > src/cpp/RiderLink/Source/RD/win-vs17
JetBrains.RdFramework.Cpp.201.1.56-win-vs19.nupkg/native > src/cpp/RiderLink/Source/RD/win-vs19

NB: all this is automated and works as part of CI for UnrealLink plugin, but UnrealLink plugin now depends on new version of Rider that hasn't been published yet. I'm waiting for team to finish polishing so I'll be able to ship plugin as well

**Comment by DecoyRS** [comment]
Hi @hikingyoung
Is your project buildable on it's own?
In your logs, you have:
`D:/UE4/UE_4.25/Engine/Plugins/Marketplace/RuntimeTransformer/Source/RuntimeTransformer/Public/Gizmos/BaseGizmo.h(41) : LogCompile: Error: An explicit Category specifier is required for Blueprint accessible functions in an Engine module.`
Looks like `RuntimeTransformer` is failing to build, not `RiderLink`.
Why UnrealBuildTool decides to build RuntimeTransformer when it builds RiderLink is an interesting question and I don't have an answer yet)

**Comment by DecoyRS** [comment]
@RVillani
Update to my previous message. I've finally finished building plugin with GitHub actions.
Although latest build is not compatible with Rider 2020.1.3, you can unzip UnrealLink, navigate to RD and copy it to your version of RiderLink.
example:
Link to the latest build of UnrealLink:
https://github.com/JetBrains/UnrealLink/suites/831180673/artifacts/9280844
How to find RiderLink in UnrealLink:
UnrealLink-2020.1.14.zip\UnrealLink\EditorPlugin\RiderLink.zip\Source\RD\

**Comment by hikingyoung** [comment]
> Hi @hikingyoung
> Is your project buildable on it's own?
> In your logs, you have:
> `D:/UE4/UE_4.25/Engine/Plugins/Marketplace/RuntimeTransformer/Source/RuntimeTransformer/Public/Gizmos/BaseGizmo.h(41) : LogCompile: Error: An explicit Category specifier is required for Blueprint accessible functions in an Engine module.`
> Looks like `RuntimeTransformer` is failing to build, not `RiderLink`.
> Why UnrealBuildTool decides to build RuntimeTransformer when it builds RiderLink is an interesting question and I don't have an answer yet)

Yes ! That's the point! The RiderLink is successfully installed after I removed the "RuntimeTransformer" from UE4 MarketPlace.
(The RuntimeTransformer is disabled by default, I just gived up to figure out the FAILED reason)
Thank you for your help and sorry for the waste of your time.

**Comment by RVillani** [comment]
@DecoyRS 
That's amazing support, Alexander! I feel like talking a work colleague :D

> latest build is not compatible with Rider 2020.1.3

So even if I did all the steps you described, Rider would stop telling me the RiderLink plugin is required but it still wouldn't be able to read variable usages on Blueprints ect, right?

**Comment by DecoyRS** [comment]
Thank you for kind words @RVillani !)
---------------------------
tl;dr Don't install latest UnrealLink to Rider, it won't work. But! You can take only part of the plugin, RD folder and copy-paste it in place of the old version of RD.
---------------------------
Long detailed answer :)
There's a number of things to clear up.
We have 3 parts:
* UnrealLink - this is a plugin for "Rider for Unreal Engine". It uses "Rider"'s plugin SDK and API. Latest build is not compatible with Rider 2020.1.3, because it uses Rider 2020.2.1 EAP SDK.
* RiderLink - is a plugin for Unreal Editor. It is packed inside UnrealLink and it's being installed when you press "install plugin in the popup". It connects with UnrealLink through JetBrains RD IPC protocol (https://github.com/JetBrains/rd). As long as DSL description of data that is being transfered through RD stays the same, UnrealLink and RiderLink are compatible with each other.
* JetBrains RD - as I've mentioned, it's a framework for IPC connection. "Rider for Unreal Editor" uses .NET version, "RiderLink" inside Unreal Editor uses C++ version. As long as ABI in RD C++ remains the same, you can safely take pre-built binaries for RD from latest plugin an use them in old RiderLink. Right now ABI is compatible, the only thing that has changed is extra folder for binaries built with VS 2019.
So my suggestion is to take only pre-built binaries and slightly modified RD.Build.cs files and put them in the old version of plugin.

**Comment by DecoyRS** [comment]
So I've fixed a ton of problems with installing RiderLink plugin, including:
* Not finding Unreal Engine installation in winreg. Use magic files from Epic Games with install info from %USER% folder in the first place, than dropback to winreg if it fails;
* Problems with linking RD generated classes with RiderLink plugin when unity builds disabled and/or shared pch are disabled
  * NB: when both non-unity and no shared pch are set, there are still some problems. I've forced all modules in RiderLink to use unity builds all the time.
Available in Rider for Unreal Engine 2020.2.1 (To be released this week)
https://github.com/JetBrains/UnrealLink/releases/tag/v2020.2.36
